### PR TITLE
Adding Ando-Li-Mathias

### PR DESCRIPTION
**Pull request by sylvchev**
This is a proposition to integrate the Ando-Li-Mathias geometric mean [1] with an implementation inspired by the Matlab toolbox of Dario Bini and Bruno Iannazzo [2].

This mean is clearly not computationally efficient but is interesting for comparison purpose, as the mean value is obtained from the recursive application of the geodesic between the point of the considered set. 

[1] T. Ando, C.-K. Li and R. Mathias, Geometric means, Linear Algebra Appl., 385-1 (2004) pp. 305-334.
[2] http://bezout.dm.unipi.it/software/mmtoolbox/

### CODE REVIEW

File: pyriemann/utils/mean.py [hunk 1]
```diff
@@ -300,6 +301,79 @@ def mean_ale(covmats, tol=10e-7, maxiter=50, sample_weight=None):
     return C
 
 
+def _sharp(A, B, t):
```
Comment (alexandrebarachant) [inline pyriemann/utils/mean.py]: There is a geodesic function for that :) it is in pyriemann.utils.geodesic

File: pyriemann/utils/mean.py [hunk 2]
```diff
@@ -300,6 +301,79 @@ def mean_ale(covmats, tol=10e-7, maxiter=50, sample_weight=None):
...
+
+    :returns: Karcher mean covariance matrix
+
+    References
+    ----------
+    [1] T. Ando, C.-K. Li and R. Mathias, "Geometric Means", Linear Algebra
+    Appl. 385 (2004), 305-334.
+    """
+    sample_weight = _get_sample_weight(sample_weight, covmats)
+    C = covmats.copy()
```
Comment (alexandrebarachant) [inline pyriemann/utils/mean.py]: Why did you copy this array ?

File: pyriemann/utils/mean.py [hunk 3]
```diff
@@ -300,6 +301,79 @@ def mean_ale(covmats, tol=10e-7, maxiter=50, sample_weight=None):
...
+    C_iter = numpy.zeros_like(C)
+    Nt, Ne, Ne = covmats.shape
+    if Nt == 2:
+        X = _sharp(covmats[0], covmats[1], 0.5)
+        return X
+    else:
+        for k in range(maxiter):
+            for h in range(Nt):
+                s = numpy.mod(numpy.arange(h, h+Nt-1)+1, Nt)
```
Comment (alexandrebarachant) [inline pyriemann/utils/mean.py]: pep8 : space bewteen arithmetic operators

File: pyriemann/utils/mean.py [hunk 4]
```diff
@@ -300,6 +301,79 @@ def mean_ale(covmats, tol=10e-7, maxiter=50, sample_weight=None):
...
+                C_iter[h, :, :] = mean_alm(C[s])
+
+            ni=numpy.linalg.norm(C_iter[0]-C[0], 2)/numpy.linalg.norm(C[0], 2)
```
Comment (alexandrebarachant) [inline pyriemann/utils/mean.py]: pep8 : space between operators

**Comment by agramfort** [comment]
@sylvchev can you update the API page https://pyriemann.readthedocs.io/en/latest/api.html and add a what's new entry? thanks @sylvchev !

File: pyriemann/utils/mean.py [hunk 5]
```diff
@@ -304,6 +305,56 @@ def mean_ale(covmats, tol=10e-7, maxiter=50, sample_weight=None):
...
+    :param sample_weight: the weight of each sample
+
+    :returns: Karcher mean covariance matrix
+
+    References
+    ----------
+    [1] T. Ando, C.-K. Li and R. Mathias, "Geometric Means", Linear Algebra
+        Appl. 385 (2004), 305-334.
+    """
+    sample_weight = _get_sample_weight(sample_weight, covmats)
```
Comment (qbarthelemy) [inline pyriemann/utils/mean.py:336]: Major problem : in the current implementation, weights are totally unused.

Comment (sylvchev) [reply pyriemann/utils/mean.py:336]: Right! It is corrected, I added weights in the computation.

Comment (qbarthelemy) [reply pyriemann/utils/mean.py:336]: Great! Can you add weights in tests too?

File: pyriemann/utils/mean.py [hunk 6]
```diff
@@ -304,6 +305,63 @@ def mean_ale(covmats, tol=10e-7, maxiter=50, sample_weight=None):
...
+        \mathbf{C} = A^{\frac{1}{2}}(A^{-\frac{1}{2}}B^{\frac{1}{2}}A^{-\frac{1}{2}})^{\frac{1}{2}}A^{\frac{1}{2}}
+
+    require a high number of iterations.
+
+    This is the adaptation of the Matlab code proposed by Dario Bini and
+    Bruno Iannazzo, http://bezout.dm.unipi.it/software/mmtoolbox/
+    Extremely slow, due to the recursive formulation.
+
+    :param covmats: Covariance matrices set, (n_trials, n_channels, n_channels)
+    :param tol: the tolerance to stop the gradient descent
```
Comment (qbarthelemy) [inline pyriemann/utils/mean.py:324]: Is it really a "gradient descent" algorithm?

Comment (sylvchev) [reply pyriemann/utils/mean.py:324]: Copy paste error, there is no gradient descent, this is the main point of the approach.

File: pyriemann/utils/mean.py [hunk 7]
```diff
@@ -40,7 +41,7 @@ def mean_riemann(covmats, tol=10e-9, maxiter=50, init=None,
     """  # noqa
     # init
     sample_weight = _get_sample_weight(sample_weight, covmats)
-    Nt, Ne, Ne = covmats.shape
+    n_trials, n_channels, n_channels = covmats.shape
```
Comment (qbarthelemy) [inline pyriemann/utils/mean.py:44]:
```suggestion
    n_trials, n_channels, _= covmats.shape
```
avoids variable redefinition.

Comment (sylvchev) [reply pyriemann/utils/mean.py:44]: Yes, I wondered about that one, but I found it more readable by explicitely indicating that they are spatial covariance matrices. And we are still below the 80-ish characters.

Comment (qbarthelemy) [reply pyriemann/utils/mean.py:44]: Yes, but complete shape is given in the docstring.

File: tests/test_utils_mean.py [hunk 8]
```diff
@@ -1,134 +1,125 @@
...
-    """Test the riemannian mean with init"""
-    covmats, diags, A = generate_cov(100, 3)
-    C = mean_riemann(covmats, init=covmats[0])
-    Ctrue = np.exp(np.log(diags).mean(0))
-    Ctrue = np.dot(np.dot(A, np.diag(Ctrue)), A.T)
-    assert_array_almost_equal(C, Ctrue)
+@pytest.mark.parametrize("mean", [mean_riemann, mean_logdet])
+def test_mean_shape_with_init(mean):
+    """Test the shape of mean with init"""
+    covmats, _, A = generate_cov(5, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:56]:
```suggestion
    n_trials, n_channels = 5, 3
    covmats, _, A = generate_cov(n_trials, n_channels)
```

File: tests/test_utils_mean.py [hunk 9]
```diff
@@ -1,134 +1,125 @@
...
+    C = mean(covmats, init=covmats[0])
+    assert C.shape == (3, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:58]:
```suggestion
    assert C.shape == (n_channels, n_channels)
```

File: tests/test_utils_mean.py [hunk 10]
```diff
@@ -1,134 +1,125 @@
...
 
-def test_logeuclid_mean():
-    """Test the logeuclidean mean"""
-    covmats, _, A = generate_cov(100, 3)
-    C = mean_logeuclid(covmats)
-    assert C.shape == (3, 3)
+@pytest.mark.parametrize("init", [True, False])
+def test_riemann_mean(init):
+    """Test the riemannian mean"""
+    covmats, diags, A = generate_cov(100, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:64]:
```suggestion
    n_trials, n_channels = 100, 3
    covmats, diags, A = generate_cov( n_trials, n_channels)
```

**Comment by sylvchev** [comment]
Thanks Quentin. I'll make a global pass on the test to use fixture and avoi code duplication, see PR #135

File: tests/test_utils_mean.py [hunk 11]
```diff
@@ -1,134 +1,127 @@
...
     """Test the euclidean mean"""
     covmats, _, _ = generate_cov(100, 3)
     C = mean_euclid(covmats)
-    assert C.shape == (3, 3)
-    assert_array_almost_equal(C, covmats.mean(axis=0))
+    assert C == approx(covmats.mean(axis=0))
 
 
 def test_identity_mean():
     """Test the logdet mean"""
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:84]: Copy-paste error in the docstring?

File: tests/test_utils_mean.py [hunk 12]
```diff
@@ -1,134 +1,127 @@
...
 
-def test_kullback_mean():
-    """Test the kullback mean"""
-    covmats, _, _ = generate_cov(100, 3)
-    C = mean_kullback_sym(covmats)
+def test_alm_mean_maxiter():
+    """Test the ALM mean with max iteration"""
+    covmats, _, _ = generate_cov(3, 3)
+    C = mean_alm(covmats, maxiter=1, verbose=True)  # maxiter reached
     assert C.shape == (3, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:102]: Shape is already tested in `test_mean_shape`, isn't it?

File: tests/test_utils_mean.py [hunk 13]
```diff
@@ -1,134 +1,127 @@
...
+
+
+@pytest.mark.parametrize(
+    "metric,mean",
+    [
+        ("riemann", mean_riemann),
+        ("logdet", mean_logdet),
+        ("logeuclid", mean_logeuclid),
+        ("euclid", mean_euclid),
+        ("alm", mean_alm),
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:119]: Why other metrics/means ('identity', 'wasserstein', 'ale', 'harmonic', 'kullback_sym') are not tested here?

File: tests/test_utils_mean.py [hunk 14]
```diff
@@ -1,134 +1,127 @@
...
+    else:
+        C = mean_riemann(covmats)
+    Ctrue = np.exp(np.log(diags).mean(0))
+    Ctrue = A @ np.diag(Ctrue) @ A.T
+    assert C == approx(Ctrue)
 
 
 def test_euclid_mean():
     """Test the euclidean mean"""
     covmats, _, _ = generate_cov(100, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:78]:
```suggestion
    n_trials, n_channels = 100, 3
    covmats, _, _ = generate_cov(n_trials, n_channels)
```

File: tests/test_utils_mean.py [hunk 15]
```diff
@@ -1,134 +1,127 @@
...
+        mean_logdet,
+        mean_ale,
+        mean_kullback_sym,
+        mean_harmonic,
+        mean_wasserstein,
+    ],
+)
+def test_mean_shape(mean):
+    """Test the shape of mean"""
+    covmats, _, A = generate_cov(5, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:48]:
```suggestion
    n_trials, n_channels = 5, 3
    covmats, _, A = generate_cov(n_trials, n_channels)
```

File: tests/test_utils_mean.py [hunk 16]
```diff
@@ -1,134 +1,129 @@
...
+        mean_harmonic,
+        mean_wasserstein,
+    ],
+)
+def test_mean_shape(mean):
+    """Test the shape of mean"""
+    n_trials, n_channels = 5, 3
+    covmats, _, A = generate_cov(n_trials, n_channels)
+    C = mean(covmats)
+    assert C.shape == (3, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:51]:
```suggestion
    assert C.shape == (n_channels , n_channels)
```

File: tests/test_utils_mean.py [hunk 17]
```diff
@@ -1,134 +1,129 @@
...
+    covmats, _, _ = generate_cov(n_trials, n_channels)
     C = mean_euclid(covmats)
-    assert C.shape == (3, 3)
-    assert_array_almost_equal(C, covmats.mean(axis=0))
+    assert C == approx(covmats.mean(axis=0))
 
 
 def test_identity_mean():
     """Test the logdet mean"""
     covmats, _, _ = generate_cov(100, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:87]:
```suggestion
    n_trials, n_channels = 100, 3
    covmats, _, _ = generate_cov(n_trials, n_channels)
```

File: tests/test_utils_mean.py [hunk 18]
```diff
@@ -1,134 +1,129 @@
...
     C = mean_identity(covmats)
-    assert C.shape == (3, 3)
-    assert_array_equal(C, np.eye(3))
+    assert np.all(C == np.eye(3))
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:89]:
```suggestion
    assert np.all(C == np.eye(n_channels))
```

File: tests/test_utils_mean.py [hunk 19]
```diff
@@ -1,134 +1,129 @@
...
 
 
-def test_logdet_mean():
-    """Test the logdet mean"""
-    covmats, _, _ = generate_cov(100, 3)
-    C = mean_logdet(covmats)
-    assert C.shape == (3, 3)
+def test_alm_mean():
+    """Test the ALM mean"""
+    covmats, _, _ = generate_cov(3, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:94]:
```suggestion
    n_trials, n_channels = 5, 3
    covmats, _, _ = generate_cov(n_trials, n_channels)
```

File: tests/test_utils_mean.py [hunk 20]
```diff
@@ -1,134 +1,129 @@
...
-    assert C.shape == (3, 3)
-
-
-def test_ald_mean():
-    """Test the Ale mean"""
-    covmats, _, _ = generate_cov(100, 3)
-    C = mean_ale(covmats)
+def test_alm_mean_maxiter():
+    """Test the ALM mean with max iteration"""
+    covmats, _, _ = generate_cov(3, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:102]:
```suggestion
    n_trials, n_channels = 5, 3
    covmats, _, _ = generate_cov(n_trials, n_channels)
```

File: tests/test_utils_mean.py [hunk 21]
```diff
@@ -1,134 +1,129 @@
...
-
-def test_mean_covariance_euclid():
-    """Test mean_covariance for euclidean metric"""
-    covmats, diags, A = generate_cov(100, 3)
-    C = mean_covariance(covmats, metric='euclid')
-    Ctrue = mean_euclid(covmats)
-    assert_array_equal(C, Ctrue)
+def test_alm_mean_2trials():
+    """Test the ALM mean with 2 trials"""
+    covmats, _, _ = generate_cov(2, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:109]:
```suggestion
    n_trials, n_channels = 2, 3
    covmats, _, _ = generate_cov(n_trials, n_channels)
```

File: tests/test_utils_mean.py [hunk 22]
```diff
@@ -1,134 +1,129 @@
...
+        ("riemann", mean_riemann),
+        ("logdet", mean_logdet),
+        ("logeuclid", mean_logeuclid),
+        ("euclid", mean_euclid),
+        ("alm", mean_alm),
+    ],
+)
+def test_mean_covariance_metric(metric, mean):
+    """Test mean_covariance for metric"""
+    covmats, _, _ = generate_cov(3, 3)
```
Comment (qbarthelemy) [inline tests/test_utils_mean.py:126]:
```suggestion
    n_trials, n_channels = 5, 3
    covmats, _, _ = generate_cov(n_trials, n_channels)
```

File: tests/test_utils_mean.py (hunk 13 above)
Comment (sylvchev) [reply tests/test_utils_mean.py:121]: yes, I just refactored existing tests, but we should tests these ones also!

File: tests/test_utils_mean.py (hunk 12 above)
Comment (sylvchev) [reply tests/test_utils_mean.py:107]: not with maxiter set

**Comment by qbarthelemy** [comment]
Ok for me. Great job!
But the branch should be rebased before merging.

**Comment by agramfort** [comment]
thx @sylvchev 

no need to rebase as we squash on merge

**Comment by sylvchev** [comment]
Thanks!

**Pull request merged**